app/controllers/api/verify/base_controller.rb

```diff
@@ -0,0 +1,25 @@
+module Api
+  module Verify
+    class BaseController < ApplicationController
+      include RenderConditionConcern
+
+      check_or_render_not_found -> do
+        IdentityConfig.store.idv_api_enabled_steps.include?(required_step)
+      end
+      before_action :confirm_two_factor_authenticated_for_api
+
+      respond_to :json
+
+      private
+
+      def confirm_two_factor_authenticated_for_api
+        return if user_fully_authenticated?
+        render json: { error: 'user is not fully authenticated' }, status: :unauthorized
+      end
+
+      def required_step
+        raise NotImplementedError
+      end
```

Comment:
Yeah, good call. At first I made it a constant, 'til I started running into the issues which led to the changes in 7dc0618. Now that I have that sorted, maybe a constant would be the better option?

We'd lose the `NotImplementedError` exception raising, but as long as the condition block would fall back to 404 if the step was missing, I think that'd be sufficient enough to force the developer to implement it.

Reply:
> I don't like dynamically accessing constant, I prefer class methods, but otherwise same same

Can you elaborate on what you mean by dynamically accessing? This is what I had in mind for implementation, which wouldn't be much different from if it were a class property.

```rb
      check_or_render_not_found -> do
        if !defined?(self.class::REQUIRED_STEP)
          raise NotImplementedError, 'Controller must define REQUIRED_STEP constant'
        end

        IdentityConfig.store.idv_api_enabled_steps.include?(self.class::REQUIRED_STEP)
      end
```

My thinking with a constant is that we'd want to communicate that it's a value that's not expected to change.

Reply:
Yeah dynamically accessing a constant with `self.class::REQUIRED_STEP`, that's the bit I dislike. I feel like in terms of expecting an interface from an object (even if the object is a Class), I expect to call methods. I much prefer `self.class.property_name` ... I was trying to find a blog post, I feel like it's one of the things Ruby encourages with its lack of distinction between direct properties and getter methods.

I feel like if I'm accessing a constant, it had better be there, and I'm used to using a full name `Some::ClassName::CONSTANT` or `CONSTANT` inside a `Some::ClassName` namespace that gets there.

Anyways, it accomplishes the same goal, and we do use dynamic constant access in FlowStateMachine, so the pattern exists, and this wouldn't be out of place in our codebase.

Reply:
It makes sense now that you explain it. Happy to update 👍

Reply:
Updated in e9d75b2 . Let me know if that's what you had in mind.